File: sycl/sycl-rs-sys/src/usm.cpp
//
// Copyright (C) 2026 Intel Corporation
//
// Under the MIT License or the Apache License v2.0.
// See LICENSE-MIT and LICENSE-APACHE for license information.
// SPDX-License-Identifier: MIT OR Apache-2.0
//

#include "sycl-rs-sys/include/usm.hpp"
#include "sycl-rs-sys/src/usm-sys.rs.h"

namespace sycl_shims::usm {
std::uint8_t *aligned_alloc_device(std::size_t alignment, std::size_t num_bytes,
                                   Queue const &queue) {
  return static_cast<std::uint8_t *>(
      sycl::aligned_alloc_device(alignment, num_bytes, queue));
}

std::uint8_t *aligned_alloc_host(std::size_t alignment, std::size_t num_bytes,
                                 Queue const &queue) {
  return static_cast<std::uint8_t *>(
      sycl::aligned_alloc_host(alignment, num_bytes, queue));
}

std::uint8_t *aligned_alloc_shared(std::size_t alignment, std::size_t num_bytes,
                                   Queue const &queue) {
  return static_cast<std::uint8_t *>(
      sycl::aligned_alloc_shared(alignment, num_bytes, queue));
}

void free(std::uint8_t *memory, Queue const &queue) {
  sycl::free(static_cast<void *>(memory), queue);
}
} // namespace sycl_shims::usm

File: sycl/sycl-rs/src/usm.rs
//
// Copyright (C) 2026 Intel Corporation
//
// Under the MIT License or the Apache License v2.0.
// See LICENSE-MIT and LICENSE-APACHE for license information.
// SPDX-License-Identifier: MIT OR Apache-2.0
//

use crate::queue::Queue;

use allocator_api2::alloc::{AllocError, Allocator};
use sycl_rs_sys::usm::ffi;

use std::{alloc::Layout, marker::PhantomData, ptr::NonNull};

type CxxResult<T> = cxx::core::result::Result<T, cxx::Exception>;

/// An instance of a USM allocator.
pub struct UsmAllocator<T: UsmAllocatorKind> {
    queue: Queue,
    _kind: PhantomData<T>,
}

/// A marker trait for USM allocators.
pub unsafe trait UsmAlloc: Allocator {}

unsafe impl<T: UsmAllocatorKind> UsmAlloc for UsmAllocator<T> {}

pub trait UsmAllocatorKind {
    unsafe fn alloc(alignment: usize, num_bytes: usize, queue: &Queue) -> CxxResult<*mut u8>;
}

/// A marker trait for host-accessible USM allocators.
pub unsafe trait HostAccessible {}

impl<T: UsmAllocatorKind> From<&Queue> for UsmAllocator<T> {
    fn from(queue: &Queue) -> Self {
        Self {
            queue: queue.clone(),
            _kind: PhantomData,
        }
    }
}

unsafe impl<T: UsmAllocatorKind> Allocator for UsmAllocator<T> {
    fn allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
        let ptr = unsafe { T::alloc(layout.align(), layout.size(), &self.queue) }
            .map_err(|_e| AllocError)?;

        let ptr = NonNull::new(ptr).ok_or(AllocError)?;
        let slice = NonNull::slice_from_raw_parts(ptr, layout.size());

        Ok(slice)
    }

    unsafe fn deallocate(&self, ptr: NonNull<u8>, _layout: Layout) {
        unsafe {
            ffi::free(ptr.as_ptr(), &self.queue.0);
        }
    }
}

/// An allocator for Device-side buffers
/// Safety: memory allocated by this allocator cannot be accessed on the host side
#[allow(dead_code)]
pub struct DeviceAllocator;

impl UsmAllocatorKind for DeviceAllocator {
    unsafe fn alloc(alignment: usize, num_bytes: usize, queue: &Queue) -> CxxResult<*mut u8> {
        unsafe { ffi::aligned_alloc_device(alignment, num_bytes, &queue.0) }
    }
}

/// An allocator for Host-side buffers
pub struct HostAllocator;

impl UsmAllocatorKind for HostAllocator {
    unsafe fn alloc(alignment: usize, num_bytes: usize, queue: &Queue) -> CxxResult<*mut u8> {
        unsafe { ffi::aligned_alloc_host(alignment, num_bytes, &queue.0) }
    }
}

unsafe impl HostAccessible for UsmAllocator<HostAllocator> {}

/// An allocator for shared memory buffers
pub struct SharedAllocator;

impl UsmAllocatorKind for SharedAllocator {
    unsafe fn alloc(alignment: usize, num_bytes: usize, queue: &Queue) -> CxxResult<*mut u8> {
        unsafe { ffi::aligned_alloc_shared(alignment, num_bytes, &queue.0) }
    }
}

unsafe impl HostAccessible for UsmAllocator<SharedAllocator> {}
